### PR TITLE
[Deck] Add Deck CRUD Operations

### LINKED ISSUE #2844: [Deck] Handle the card selection and direct edit in the Deck representation
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

- [x] When selecting a card, the target semantic element is selected.
- [x] When selecting the card target semantic element, the card is selected.
- [x] When a card is selected, it can be edit by a second click on the label, description, etc.
- [x] When a card is selected,pressing F2 trigger the label editing. When pressing "Tab", the next field get the focus (label, description)

...

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### LINKED ISSUE #2871: [Deck] Handle Edit, Add and Remove Card operations
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

- [x] An existing card can be modified if the edit tool exists
- [x] An existing card can be deleted if the delete tool exists
- [x] A new card can be created if the creation tool exists

...

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by florianbarbin**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?

- Create a new Deck view model from the provided stereotype
- Import the [test.zip](https://github.com/eclipse-sirius/sirius-web/files/13857934/test.zip)
- Open the Deck representation
- Try to Edit Card
- Try to Add a Card
- Try to remove a card
- A card should be editable once selected. Once selected, press F2 or click to start editing

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+export const cardDetailFontStyle: CSSProperties = {
+  fontSize: '12px',
+};
+
+export const DeckDraggableCardWrapper = styled('article')({
+  borderRadius: ' 3px',
+  borderBottom: '1px solid rgb(204, 204, 204)',
+  backgroundColor: 'rgb(255, 255, 255)',
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx]: We avoid using custom color, we prefer to use theme color

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  padding: '0',
+  textAlign: 'center',
+  width: '15px',
+  background: 'inherit',
+  cursor: 'pointer',
+  opacity: '1',
+});
+
+export const Detail = styled('div')({
+  color: '#4d4d4d',
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx]: avoid custom color

File: packages/deck/frontend/sirius-components-deck/src/card/DeckDeleteButton.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+import { DelButton, DeleteWrapper } from './DeckCardStyledComponents';
+
+export const DeckDeleteButton = (props) => {
+  return (
+    <DeleteWrapper {...props}>
+      <DelButton>&#10006;</DelButton>
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckDeleteButton.tsx]: we usually use a trash icon in place of x

File: packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx [hunk 4]
```diff
@@ -83,6 +91,36 @@ export const DeckRepresentation = ({ editingContextId, representationId }: Repre
...
+    }
+  }, [selection]);
+
+  const handleCardClicked = (cardId: string, metadata: CardMetadata, _laneId: string) => {
+    let editedCardId: string | undefined = undefined;
+    if (selectedCardIds.includes(cardId)) {
+      editedCardId = cardId;
+    }
+    setState((prevState) => {
+      return { ...prevState, editedCardId };
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx]: `editedCardId` state seems never used

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  const StyledTextField = styled(TextField)({
+    '& .MuiInputBase-multiline': {
+      padding: '0px 0px 0px 0px',
+    },
+    '& input': {
+      padding: '0px 0px 0px 0px',
+    },
+    '& input:focus': {
+      backgroundColor: 'white',
+      boxShadow: 'inset 0 0 0 2px #0079bf',
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx]: use them color

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    },
+    '& textarea': {
+      padding: '0px 0px 0px 0px',
+    },
+    '& textarea:focus': {
+      backgroundColor: 'white',
+      boxShadow: 'inset 0 0 0 2px #0079bf',
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx]: use theme color

File: packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx [hunk 7]
```diff
@@ -77,12 +99,127 @@ export const DeckRepresentation = ({ editingContextId, representationId }: Repre
...
+            tempSelectedCardIds.push(card.id);
+          }
+        });
+      setState((prevState) => {
+        return { ...prevState, selectedCardIds: tempSelectedCardIds };
+      });
+    }
+  }, [selection]);
+
+  const isStandardErrorPayload = (field): field is GQLErrorPayload => field.__typename === 'ErrorPayload';
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx]: could be extract from component like `isErrorPayload`

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+ * Service used to manage cards.
+ *
+ * @author fbarbin
+ */
+@Service
+public class DeckCardService implements IDeckCardService {
+
+    private final IRepresentationDescriptionSearchService representationDescriptionSearchService;
+
+    private IFeedbackMessageService feedbackMessageService;
```
Comment (frouene) [inline packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: may be final

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        if (optionalErrorMsg.isPresent()) {
+            payload = new ErrorPayload(payloadId, optionalErrorMsg.get().body(), feedbackMessages);
+        } else {
+            payload = new SuccessPayload(payloadId, feedbackMessages);
+        }
+        return payload;
+    }
+    private Optional<Lane> findLane(Predicate<Lane> condition, Deck deck) {
+        return deck.lanes().stream()
+                .filter(condition::test)
```
Comment (frouene) [inline packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: condition is already a predicate, you don't have to call test you could just use `.filter(condition)`

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                .findFirst();
+    }
+
+    private Optional<Card> findCard(Predicate<Card> condition, Deck deck) {
+        return deck.lanes().stream()
+                .flatMap(lane -> lane.cards().stream())
+                .filter(condition::test)
```
Comment (frouene) [inline packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: same remark

File: package-lock.json [hunk 11]
```diff
@@ -6617,17 +6618,18 @@
       "devDependencies": {
         "@apollo/client": "3.8.1",
         "@eclipse-sirius/sirius-components-core": "*",
+        "@eclipse-sirius/sirius-components-selection": "*",
```
Comment (frouene) [inline package-lock.json:6621]: Are you sur you need the dependency on components-selection for deck ?
I see gantt have this dependency too, but it's not related to the explorer selection, it's a representation use to display a modal.

**Review by frouene — CHANGES_REQUESTED** [review]

**Comment by frouene** [comment]
There are few errors in the browser console from react-trello

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx (hunk 1 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx (hunk 2 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckCardStyledComponents.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx (hunk 4 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx]: Done

Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/card/DeckDeleteButton.tsx (hunk 3 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckDeleteButton.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx (hunk 5 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx (hunk 6 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckCardInput.tsx]: done

File: packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx (hunk 7 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/representation/DeckRepresentation.tsx]: done

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java (hunk 8 above)
Comment (florianbarbin) [reply packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: done

File: package-lock.json (hunk 11 above)
Comment (florianbarbin) [reply package-lock.json:6621]: Removed from both deck and gantt. It is a mistake, this dependencies has nothing to do here.

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java (hunk 9 above)
Comment (florianbarbin) [reply packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: done

File: packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java (hunk 10 above)
Comment (florianbarbin) [reply packages/deck/backend/sirius-components-collaborative-deck/src/main/java/org/eclipse/sirius/components/collaborative/deck/service/DeckCardService.java]: done

**Review by frouene — APPROVED** [review]

**Comment by frouene** [comment]
Falling test is not related to the PR

**Pull request merged**